core/services/workflows/store/store_memory.go

```diff
@@ -0,0 +1,189 @@
+	"time"
+
+	"github.com/jonboulle/clockwork"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	commonservices "github.com/smartcontractkit/chainlink-common/pkg/services"
+)
+
+const (
+	// defaultPruneInterval is the default interval between pruning completed executions
+	defaultPruneInterval = 30 * time.Second
+
+	// maximumExecutionAge is the default maximum age of an execution before it is considered expired and eligible for pruning
+	// regardless of its status
+	maximumExecutionAge = 1 * time.Hour
+)
+
+// InMemoryStore is an in-memory implementation of the Store interface used to store workflow execution states
+type InMemoryStore struct {
+	lggr logger.Logger
+	commonservices.StateMachine
+	idToState         map[string]*WorkflowExecution
+	mu                sync.RWMutex
+	shutdownWaitGroup sync.WaitGroup
+	chStop            commonservices.StopChan
+
+	clock clockwork.Clock
+
+	// pruneInterval is the interval between pruning completed (and expired) executions
+	pruneInterval time.Duration
+
+	// maximumExecutionAge is the maximum age of an execution before it is considered expired and eligible for pruning
+	// regardless of its status
+	maximumExecutionAge time.Duration
+}
+
+func NewInMemoryStore(lggr logger.Logger, clock clockwork.Clock) *InMemoryStore {
+	return NewInMemoryStoreWithPruneConfiguration(lggr, clock, defaultPruneInterval, maximumExecutionAge)
+}
+
+func NewInMemoryStoreWithPruneConfiguration(lggr logger.Logger, clock clockwork.Clock, pruneFrequency time.Duration,
+	maximumExecutionAge time.Duration) *InMemoryStore {
+	return &InMemoryStore{lggr: lggr, idToState: map[string]*WorkflowExecution{}, clock: clock, chStop: make(chan struct{}),
+		pruneInterval: pruneFrequency, maximumExecutionAge: maximumExecutionAge}
+}
+
+// Add adds a new execution state under the given executionID
+func (s *InMemoryStore) Add(ctx context.Context, execution *WorkflowExecution) (WorkflowExecution, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	_, ok := s.idToState[execution.ExecutionID]
+	if ok {
+		return WorkflowExecution{}, fmt.Errorf("execution ID %s already exists in store", execution.ExecutionID)
+	}
+
+	now := s.clock.Now()
+	execution.CreatedAt = &now
+	execution.UpdatedAt = &now
+
+	s.idToState[execution.ExecutionID] = execution
+	return *execution, nil
+}
+
+// UpsertStep updates a step for the given executionID
+func (s *InMemoryStore) UpsertStep(ctx context.Context, step *WorkflowExecutionStep) (WorkflowExecution, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	execution, ok := s.idToState[step.ExecutionID]
+	if !ok {
+		return WorkflowExecution{}, fmt.Errorf("could not find execution %s", step.ExecutionID)
+	}
+
+	now := s.clock.Now()
+	execution.UpdatedAt = &now
+
+	execution.Steps[step.Ref] = step
+	return *execution, nil
+}
+
+// UpdateStatus updates the status for the given executionID
+func (s *InMemoryStore) FinishedExecution(ctx context.Context, executionID string, status string) (WorkflowExecution, error) {
```

Comment:
Can we rename this to FinishExecution to imply that this method does the finishing :) ?